src/Microsoft.OData.Client/DataServiceContext.cs

```diff
@@ -2353,12 +2353,60 @@ public virtual DataServiceResponse DeepInsert<T>(T resource)
                 throw Error.ArgumentNull(nameof(resource));
             }
 
-            DeepInsertSaveResult result = new DeepInsertSaveResult(this, Util.SaveChangesMethodName, SaveChangesOptions.DeepInsert, callback: null, state: null);
+            DeepInsertSaveResult result = new DeepInsertSaveResult(this, Util.DeepInsertMethodName, SaveChangesOptions.DeepInsert, callback: null, state: null);
             result.DeepInsertRequest(resource);
 
             return result.EndRequest();
         }
 
+        /// <summary>
+        /// Asynchronously processes a deep insert request.
+        /// </summary>
+        /// <typeparam name="T">The type of top-level object to be deep inserted.</typeparam>
+        /// <param name="resource">The top-level object of the type to be deep inserted.</param>
+        /// <returns>A task representing the <see cref="DataServiceResponse"/> that holds the result of a bulk operation.</returns>
+        public virtual Task<DataServiceResponse> DeepInsertAsync<T>(T resource)
+        {
+            return this.DeepInsertAsync(resource, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Asynchronously processes a deep insert request.
+        /// </summary>
+        /// <typeparam name="T">The type of top-level object to be deep inserted.</typeparam>
+        /// <param name="resource">The top-level object of the type to be deep inserted.</param>
+        /// <returns>A task representing the <see cref="DataServiceResponse"/> that holds the result of a bulk operation.</returns>
+        public virtual Task<DataServiceResponse> DeepInsertAsync<T>(T resource, CancellationToken cancellationToken)
+        {
+            return FromAsync((objectsArg, callback, state) => BeginDeepInsert(callback, state, objectsArg), EndDeepInsert, resource, cancellationToken);
+        }
+
+        /// <summary>Asynchronously submits top-level objects to be deep inserted to the data service.</summary>
+        /// <param name="callback">The delegate that is called when a response to the deep insert request is received.</param>
+        /// <param name="state">User-defined state object that is used to pass context data to the callback method.</param>
+        /// <returns>An<see cref="IAsyncResult" /> object that is used to track the status of the asynchronous operation.</returns>
+        internal virtual IAsyncResult BeginDeepInsert<T>(AsyncCallback callback, object state, T resource)

```

Comment:
internal virtual? What does it mean? Only override internally? if so, what's the significance?